File: vercel-migration/scripts/fix-trial-subs.py
#!/usr/bin/env python3

"""
Script para corregir suscripciones en trial que tienen plan_tier incorrecto

Problema: Usuarios con trial de PRO pero tienen plan_tier='basico' en la BD
Resultado: Se les aplican límites de BÁSICO (2 doctores) en lugar de PRO (10 doctores)

Uso: python scripts/fix-trial-subs.py [email]
"""

import os
import sys
from supabase import create_client, Client
from dotenv import load_dotenv
from pathlib import Path

# Cargar variables de entorno desde vercel-migration/.env.local
env_path = Path(__file__).parent.parent / '.env.local'
load_dotenv(env_path)

PLAN_LIMITS = {
    'basico': {'max_doctors': 2, 'max_locations': 1},
    'pro': {'max_doctors': 10, 'max_locations': 5},
    'enterprise': {'max_doctors': 999, 'max_locations': 999}
}

def infer_plan_from_price_id(price_id):
    """Inferir el plan correcto desde el Stripe Price ID"""
    if not price_id:
        return 'basico'
    if 'pro' in price_id.lower():
        return 'pro'
    if 'enterprise' in price_id.lower():
        return 'enterprise'
    return 'basico'

def create_supabase_client() -> Client:
    """Crear cliente de Supabase con las credenciales del entorno"""
    url = os.getenv('NEXT_PUBLIC_SUPABASE_URL')
    key = os.getenv('SUPABASE_SERVICE_ROLE_KEY')
    
    if not url or not key:
        print('❌ Error: Faltan credenciales de Supabase')
        print(f'NEXT_PUBLIC_SUPABASE_URL: {"✓" if url else "✗"}')
        print(f'SUPABASE_SERVICE_ROLE_KEY: {"✓" if key else "✗"}')
        sys.exit(1)
    
    return create_client(url, key)

def list_trial_users(supabase: Client):
    """Listar todos los usuarios en trial y detectar inconsistencias"""
    print('📋 Listando todos los usuarios en trial...\n')
    
    try:
        # Obtener todas las suscripciones en trial
        response = supabase.table('subscriptions')\
            .select('*')\
            .eq('status', 'trialing')\
            .order('created_at', desc=True)\
            .execute()
        
        subscriptions = response.data
        
        if not subscriptions:
            print('✓ No hay usuarios en trial actualmente')
            return
        
        print(f'Encontrados {len(subscriptions)} usuarios en trial:\n')
        
        # Para cada suscripción, obtener el email del usuario
        for sub in subscriptions:
            user_id = sub['user_id']
            
            # Obtener usuario desde auth.users
            try:
                user_response = supabase.auth.admin.get_user_by_id(user_id)
                email = user_response.user.email if user_response and user_response.user else 'Email no disponible'
            except Exception as e:
                email = f'Error obteniendo email: {str(e)}'
            
            inferred_plan = infer_plan_from_price_id(sub.get('stripe_price_id'))
            expected_limits = PLAN_LIMITS[inferred_plan]
            has_issue = (sub['plan_tier'] != inferred_plan or 
                        sub['max_doctors'] != expected_limits['max_doctors'])
            
            icon = '⚠️ ' if has_issue else '✓ '
            print(f'{icon} {email}')
            print(f'   Plan actual: {sub["plan_tier"]} ({sub["max_doctors"]} doctores, {sub["max_locations"]} ubicaciones)')
            print(f'   Stripe Price: {sub.get("stripe_price_id", "N/A")}')
            print(f'   Plan inferido: {inferred_plan} (debería tener {expected_limits["max_doctors"]} doctores)')
            print(f'   Trial termina: {sub.get("trial_end", "N/A")}')
            if has_issue:
                print('   🔧 NECESITA CORRECCIÓN')
            print()
            
    except Exception as e:
        print(f'❌ Error: {str(e)}')
        import traceback
        traceback.print_exc()

def fix_user_trial_plan(supabase: Client, email: str):
    """Corregir la suscripción de un usuario específico"""
    print(f'🔍 Buscando usuario: {email}\n')
    
    try:
        # Buscar usuario por email
        user_response = supabase.auth.admin.list_users()
        user = None
        
        if user_response and hasattr(user_response, 'users'):
            for u in user_response.users:
                if u.email == email:
                    user = u
                    break
        
        if not user:
            print('❌ No se encontró usuario con ese email')
            return
        
        # Buscar suscripción en trial
        response = supabase.table('subscriptions')\
            .select('*')\
            .eq('user_id', user.id)\
            .eq('status', 'trialing')\
            .execute()
        
        if not response.data:
            print('❌ No se encontró suscripción en trial para este usuario')
            return
        
        sub = response.data[0]
        inferred_plan = infer_plan_from_price_id(sub.get('stripe_price_id'))
        expected_limits = PLAN_LIMITS[inferred_plan]
        
        print('📊 Situación actual:')
        print(f'   Email: {email}')
        print(f'   Plan actual: {sub["plan_tier"]}')
        print(f'   Límites actuales: {sub["max_doctors"]} doctores, {sub["max_locations"]} ubicaciones')
        print(f'   Stripe Price ID: {sub.get("stripe_price_id", "N/A")}')
        print()
        print('🎯 Plan correcto inferido:')
        print(f'   Plan: {inferred_plan}')
        print(f'   Límites correctos: {expected_limits["max_doctors"]} doctores, {expected_limits["max_locations"]} ubicaciones')
        print()
        
        if (sub['plan_tier'] == inferred_plan and 
            sub['max_doctors'] == expected_limits['max_doctors'] and
            sub['max_locations'] == expected_limits['max_locations']):
            print('✓ El usuario ya tiene el plan correcto. No se necesitan cambios.')
            return
        
        # Pedir confirmación
        answer = input('¿Actualizar suscripción? (si/no): ')
        
        if answer.lower() not in ['si', 's', 'yes', 'y']:
            print('❌ Operación cancelada')
            return
        
        # Actualizar suscripción
        from datetime import datetime
        update_response = supabase.table('subscriptions')\
            .update({
                'plan_tier': inferred_plan,
                'max_doctors': expected_limits['max_doctors'],
                'max_locations': expected_limits['max_locations'],
                'updated_at': datetime.utcnow().isoformat()
            })\
            .eq('id', sub['id'])\
            .execute()
        
        if update_response.data:
            updated_sub = update_response.data[0]
            print('\n✅ Suscripción actualizada exitosamente:')
            print(f'   Plan: {updated_sub["plan_tier"]}')
            print(f'   Límites: {updated_sub["max_doctors"]} doctores, {updated_sub["max_locations"]} ubicaciones')
        else:
            print('❌ Error al actualizar la suscripción')
            
    except Exception as e:
        print(f'❌ Error: {str(e)}')
        import traceback
        traceback.print_exc()

def main():
    """Función principal"""
    supabase = create_supabase_client()
    
    if len(sys.argv) > 1:
        email = sys.argv[1]
        fix_user_trial_plan(supabase, email)
    else:
        list_trial_users(supabase)

if __name__ == '__main__':
    main()

File: vercel-migration/scripts/fix-trial-subs.ts
/**
 * Script para corregir suscripciones en trial que tienen plan_tier incorrecto
 */

import { createClient } from '@supabase/supabase-js'
import * as dotenv from 'dotenv'
import * as path from 'path'
import * as readline from 'readline'

// Cargar .env.local desde el directorio raíz del proyecto
dotenv.config({ path: path.join(__dirname, '..', '.env.local') })

type PlanTier = 'basico' | 'pro' | 'enterprise'

interface PlanLimits {
  max_doctors: number
  max_locations: number
}

const PLAN_LIMITS: Record<PlanTier, PlanLimits> = {
  basico: { max_doctors: 2, max_locations: 1 },
  pro: { max_doctors: 10, max_locations: 5 },
  enterprise: { max_doctors: 999, max_locations: 999 }
}

function getErrorMessage(error: unknown): string {
  return error instanceof Error ? error.message : String(error)
}

function inferPlanFromPriceId(priceId: string | null): PlanTier {
  if (!priceId) return 'basico'
  if (priceId.includes('pro')) return 'pro'
  if (priceId.includes('enterprise')) return 'enterprise'
  return 'basico'
}

// Crear cliente de Supabase
const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL!
const serviceRoleKey = process.env.SUPABASE_SERVICE_ROLE_KEY!

if (!supabaseUrl || !serviceRoleKey) {
  console.error('❌ Faltan credenciales de Supabase')
  console.error('NEXT_PUBLIC_SUPABASE_URL:', supabaseUrl ? '✓' : '✗')
  console.error('SUPABASE_SERVICE_ROLE_KEY:', serviceRoleKey ? '✓' : '✗')
  process.exit(1)
}

const supabase = createClient(supabaseUrl, serviceRoleKey, {
  auth: {
    autoRefreshToken: false,
    persistSession: false
  }
})

async function listTrialUsers() {
  console.log('📋 Listando todos los usuarios en trial...\n')

  try {
    const { data: subscriptions, error } = await supabase
      .from('subscriptions')
      .select('*')
      .eq('status', 'trialing')
      .order('created_at', { ascending: false })

    if (error) {
      console.error('❌ Error al consultar suscripciones:', error.message)
      return
    }

    if (!subscriptions || subscriptions.length === 0) {
      console.log('✓ No hay usuarios en trial actualmente')
      return
    }

    console.log(`Encontrados ${subscriptions.length} usuarios en trial:\n`)

    for (const sub of subscriptions) {
      const { data: userData } = await supabase.auth.admin.getUserById(sub.user_id)
      const email = userData?.user?.email || 'Email no disponible'
      
      const inferredPlan = inferPlanFromPriceId(sub.stripe_price_id as string | null)
      const expectedLimits = PLAN_LIMITS[inferredPlan]
      const hasIssue = sub.plan_tier !== inferredPlan || 
                       sub.max_doctors !== expectedLimits.max_doctors

      console.log(`${hasIssue ? '⚠️ ' : '✓ '} ${email}`)
      console.log(`   Plan actual: ${sub.plan_tier} (${sub.max_doctors} doctores, ${sub.max_locations} ubicaciones)`)
      console.log(`   Stripe Price: ${sub.stripe_price_id || 'N/A'}`)
      console.log(`   Plan inferido: ${inferredPlan} (debería tener ${expectedLimits.max_doctors} doctores)`)
      console.log(`   Trial termina: ${sub.trial_end || 'N/A'}`)
      if (hasIssue) {
        console.log('   🔧 NECESITA CORRECCIÓN')
      }
      console.log()
    }
  } catch (error: unknown) {
    console.error('❌ Error:', getErrorMessage(error))
    console.error(error)
  }
}

async function fixUserTrialPlan(email: string) {
  console.log(`🔍 Buscando usuario: ${email}\n`)

  try {
    const { data: usersData } = await supabase.auth.admin.listUsers()
    const user = usersData?.users?.find((candidate) => candidate.email === email)

    if (!user) {
      console.log('❌ No se encontró usuario con ese email')
      return
    }

    const { data: subscriptions, error: subError } = await supabase
      .from('subscriptions')
      .select('*')
      .eq('user_id', user.id)
      .eq('status', 'trialing')

    if (subError) {
      console.error('❌ Error al buscar suscripción:', subError.message)
      return
    }

    if (!subscriptions || subscriptions.length === 0) {
      console.log('❌ No se encontró suscripción en trial para este usuario')
      return
    }

    const sub = subscriptions[0]
    const inferredPlan = inferPlanFromPriceId(sub.stripe_price_id as string | null)
    const expectedLimits = PLAN_LIMITS[inferredPlan]

    console.log('📊 Situación actual:')
    console.log(`   Email: ${email}`)
    console.log(`   Plan actual: ${sub.plan_tier}`)
    console.log(`   Límites actuales: ${sub.max_doctors} doctores, ${sub.max_locations} ubicaciones`)
    console.log(`   Stripe Price ID: ${sub.stripe_price_id || 'N/A'}`)
    console.log()
    console.log('🎯 Plan correcto inferido:')
    console.log(`   Plan: ${inferredPlan}`)
    console.log(`   Límites correctos: ${expectedLimits.max_doctors} doctores, ${expectedLimits.max_locations} ubicaciones`)
    console.log()

    if (sub.plan_tier === inferredPlan && 
        sub.max_doctors === expectedLimits.max_doctors &&
        sub.max_locations === expectedLimits.max_locations) {
      console.log('✓ El usuario ya tiene el plan correcto. No se necesitan cambios.')
      return
    }

    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout
    })

    const answer = await new Promise<string>((resolve) => {
      rl.question('¿Actualizar suscripción? (si/no): ', resolve)
    })
    rl.close()

    if (!['si', 's', 'yes', 'y'].includes(answer.toLowerCase())) {
      console.log('❌ Operación cancelada')
      return
    }

    const { data: updated, error: updateError } = await supabase
      .from('subscriptions')
      .update({
        plan_tier: inferredPlan,
        max_doctors: expectedLimits.max_doctors,
        max_locations: expectedLimits.max_locations,
        updated_at: new Date().toISOString()
      })
      .eq('id', sub.id)
      .select()

    if (updateError) {
      console.error('❌ Error al actualizar:', updateError.message)
      return
    }

    const updatedSub = updated[0]
    console.log('\n✅ Suscripción actualizada exitosamente:')
    console.log(`   Plan: ${updatedSub.plan_tier}`)
    console.log(`   Límites: ${updatedSub.max_doctors} doctores, ${updatedSub.max_locations} ubicaciones`)

  } catch (error: unknown) {
    console.error('❌ Error:', getErrorMessage(error))
    console.error(error)
  }
}

// Main
const email = process.argv[2]

if (email) {
  fixUserTrialPlan(email).then(() => process.exit(0))
} else {
  listTrialUsers().then(() => process.exit(0))
}
